core/src/main/java/io/jenkins/pluginhealth/scoring/probes/IncrementalBuildDetectionProbe.java

```diff
@@ -0,0 +1,153 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.jenkins.pluginhealth.scoring.probes;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+/**
+ * This probe checks for Incremental Build configuration in plugins
+ */
+@Component
+@Order(IncrementalBuildDetectionProbe.ORDER)
+public class IncrementalBuildDetectionProbe extends Probe {
+    public static final int ORDER = SCMLinkValidationProbe.ORDER + 100;
+    public static final String KEY = "incremental-build-maven-configuration";
+    private static final Logger LOGGER = LoggerFactory.getLogger(IncrementalBuildDetectionProbe.class);
+    private static final String INCREMENTAL_TOOL = "io.jenkins.tools.incrementals";
+    private static final String ARTIFACT_ID = "git-changelist-maven-extension";
+
+    @Override
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        final Path scmRepository = context.getScmRepository();
+        final Path mvnConfig = scmRepository.resolve(".mvn");
+        if (Files.notExists(mvnConfig)) {
+            LOGGER.info("Could not find Maven configuration folder {} plugin while running {} probe.", plugin.getName(), key());
+            return ProbeResult.failure(key(), String.format("Could not find Maven configuration folder for the %s plugin.", plugin.getName()));
+        }
+
+        try (Stream<Path> paths = Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path))) {
+            Map<String, Path> map = paths
+                .collect(Collectors.toMap(
+                    filePath -> {
+                        // building keys
+                        if (filePath.endsWith("maven.config")) {
+                            return "maven";
+                        }
+                        if (filePath.endsWith("extensions.xml")) {
+                            return "extensions";
+                        }
+                        return null;
+                    },
+                    // setting the values
+                    filePath -> filePath));
```

Comment:
That looks like an overly complex way of doing what you want to do.

Basically you want: 

* To search in `.mvn` for 2 files, `maven.config` and `extensions.xml`
* If they exist run some checks on them

Something like:

```java
try (Stream<Path> incrementalsConfigs = Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path) && (path.endsWith("maven.config") || path.endsWith("extensions.xml")))) {
```

Should directly give you a Stream that should contain 2 elements being the files you are searching for.

You can simply go for something like:

```java
Optional<Path> mavenExtensionsFile = incrementalsConfigs.filter(path -> path.endsWith("extensions.xml")).findFirst(); 

Optional<Path> mavenConfigFile = incrementalsConfigs.filter(path -> path.endsWith("maven.config")).findFirst(); 
```

Then `map` those `Optional` with whatever validation method you implemented, and if both result are present and correct, you've got a Probe Success =)

Reply:
> ```java
> Optional<Path> mavenExtensionsFile = incrementalsConfigs.filter(path -> path.endsWith("extensions.xml")).findFirst(); 
> 
> Optional<Path> mavenConfigFile = incrementalsConfigs.filter(path -> path.endsWith("maven.config")).findFirst(); 
> ```

I tried something like this initially. It throws an error saying `java.lang.IllegalStateException: stream has already been operated upon or closed`.

I read about it and learned that `Streams` are one-time use only. 

I implemented what you suggested and got the same error.  I can only execute the first statement. Not the second one :(

Reply:
One of the solutions is to use `Supplier<Stream<Path>>` but then I cannot use it in `try` because it requires auto-closeable types.

Reply:
I took a different approach and was able to handle `Supplier` this time.

Reply:
Oh yeah indeed, sorry I misled with code examples. The usual ways to tackle this are either to store things in Collection forms and generate new streams whenever needed, or indeed go for suppliers.

Suppliers seem like a good idea for your situation, and great job figuring out a solution!